In [1]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import cross_val_score, cross_validate
from sklearn.datasets import load_iris

import numpy as np

### 저번시간에 진행한 kfold skfold 더 쉽게 하는 방법
- 사이킷런에서 제공하고 있는 cross_val_score 교차검증 가능
- cross_val_score(사용할 모델, x(데이터), y(데이터), scoring(평가할 스코어링 방법), cv(몇 번 교차검증 할 것인가?), n_jobs(cpu 사용에 대한 지정, -1이면 모든 cpu를 사용), verbose(진행상황에 대한 유무 보기), fit_params 등등)
- 대부분 사용하는 건 내가 사용할 모델, 데이터, 평가점수, 몇 번 교차검증?
- cross_val_score 자동적으로 분류 or 회귀에 따라서 kfold, skfold 변경된다.
- 우리가 모델을 입력할 때 Classifier 자동으로 skf / 회귀 진행 시 KFold로 바뀐다.

- 대부분의 사이킷런 제공은 이진 분류에서 많이 사용되고, 특히 평가지표들은 대부분 이진분류로 진행된다.

In [2]:
iris_data = load_iris()
dt_clf = DecisionTreeClassifier(random_state=111)

data = iris_data.data
label = iris_data.target

In [3]:
# 교차검증 진행 (accuracy)
scores = cross_val_score(dt_clf, data, label, scoring="accuracy", cv=10)
print(np.round(scores, 4))
print(np.round(np.mean(scores), 4))

[1.     0.9333 1.     0.9333 0.9333 0.8667 0.9333 0.9333 1.     1.    ]
0.9533


In [4]:
# 교차검증 진행 (precision)
scores = cross_val_score(dt_clf, data, label, scoring="precision_macro", cv=10)
print(np.round(scores, 4))
print(np.round(np.mean(scores), 4))

[1.     0.9444 1.     0.9444 0.9444 0.8667 0.9444 0.9444 1.     1.    ]
0.9589


In [5]:
# 교차검증 진행 (recall)
scores = cross_val_score(dt_clf, data, label, scoring="recall_macro", cv=10)
print(np.round(scores, 4))
print(np.round(np.mean(scores), 4))

[1.     0.9333 1.     0.9333 0.9333 0.8667 0.9333 0.9333 1.     1.    ]
0.9533


In [6]:
# 교차검증 진행 (f1 score)
scores = cross_val_score(dt_clf, data, label, scoring="f1_macro", cv=10)
print(np.round(scores, 4))
print(np.round(np.mean(scores), 4))

[1.     0.9327 1.     0.9327 0.9327 0.8667 0.9327 0.9327 1.     1.    ]
0.953


In [7]:
label

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

### 이진 분류 진행 시?

In [8]:
import seaborn as sns
df = sns.load_dataset("titanic")

In [9]:
df

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone  
0      man        True  NaN  Southampton    no  False  
1    woman       False    C    Cherbourg   yes  False  
2    woman       False  NaN  Southampton   yes   True  
3    woman       False    C  Southampton   yes  False  
4      man        True  NaN  Southampton    no   True  
..     ...         ...  ...          ...   ...    ...  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True  

[891 rows x 15 columns]

In [10]:
df_tt = df[["survived","pclass","sibsp","parch","fare"]]

In [11]:
df_tt_x = df_tt[["pclass","sibsp","parch","fare"]]

In [12]:
df_tt_y = df["survived"]

In [13]:
df_tt_x

pclass  sibsp  parch     fare
0         3      1      0   7.2500
1         1      1      0  71.2833
2         3      0      0   7.9250
3         1      1      0  53.1000
4         3      0      0   8.0500
..      ...    ...    ...      ...
886       2      0      0  13.0000
887       1      0      0  30.0000
888       3      1      2  23.4500
889       1      0      0  30.0000
890       3      0      0   7.7500

[891 rows x 4 columns]

In [14]:
df_tt_y

0      0
1      1
2      1
3      1
4      0
      ..
886    0
887    1
888    0
889    1
890    0
Name: survived, Length: 891, dtype: int64

In [15]:
# 교차검증 진행 (accuracy)
scores = cross_val_score(dt_clf, df_tt_x, df_tt_y, scoring="accuracy", cv=10)
print(np.round(scores, 4))
print(np.round(np.mean(scores), 4))

[0.6889 0.6404 0.6742 0.6292 0.6629 0.7528 0.6629 0.7753 0.7079 0.6854]
0.688


In [16]:
# 교차검증 진행 (precision)
scores = cross_val_score(dt_clf, df_tt_x, df_tt_y, scoring="precision", cv=10)
print(np.round(scores, 4))
print(np.round(np.mean(scores), 4))

[0.6667 0.5455 0.6316 0.52   0.5833 0.6875 0.5769 0.7188 0.6429 0.6522]
0.6225


In [17]:
# 교차검증 진행 (recall)
scores = cross_val_score(dt_clf, df_tt_x, df_tt_y, scoring="recall", cv=10)
print(np.round(scores, 4))
print(np.round(np.mean(scores), 4))

[0.4    0.3529 0.3529 0.3824 0.4118 0.6471 0.4412 0.6765 0.5294 0.4286]
0.4623


In [18]:
# 교차검증 진행 (f1 score)
scores = cross_val_score(dt_clf, df_tt_x, df_tt_y, scoring="f1", cv=10)
print(np.round(scores, 4))
print(np.round(np.mean(scores), 4))

[0.5    0.4286 0.4528 0.4407 0.4828 0.6667 0.5    0.697  0.5806 0.5172]
0.5266


cross_validate - 소요시간이랑 train, test에 대한 모든 정확도 점수를 알 수 있다.

In [19]:
cross_validate(dt_clf, df_tt_x, df_tt_y, scoring=["accuracy"], cv=10, return_train_score=True)

{'fit_time': array([0.00948882, 0.00434494, 0.00804257, 0.00298238, 0.00198841,
        0.00298643, 0.00199151, 0.00298262, 0.00199485, 0.00285363]),
 'score_time': array([0.00736642, 0.00358987, 0.00199127, 0.00099683, 0.00237322,
        0.0009954 , 0.00099802, 0.00099492, 0.00099587, 0.00207305]),
 'test_accuracy': array([0.68888889, 0.64044944, 0.6741573 , 0.62921348, 0.66292135,
        0.75280899, 0.66292135, 0.7752809 , 0.70786517, 0.68539326]),
 'train_accuracy': array([0.83895131, 0.84538653, 0.84413965, 0.84538653, 0.83915212,
        0.83790524, 0.840399  , 0.83541147, 0.840399  , 0.83541147])}

In [20]:
cross_validate(dt_clf, df_tt_x, df_tt_y, scoring=["f1"], cv=10, return_train_score=True)

{'fit_time': array([0.00298214, 0.00510621, 0.00797153, 0.00405908, 0.00452995,
        0.00473261, 0.00198817, 0.0020647 , 0.00559211, 0.00298429]),
 'score_time': array([0.00208974, 0.00594497, 0.00400186, 0.00454402, 0.00099325,
        0.00298667, 0.00199366, 0.0019331 , 0.00351739, 0.00199294]),
 'test_f1': array([0.5       , 0.42857143, 0.45283019, 0.44067797, 0.48275862,
        0.66666667, 0.5       , 0.6969697 , 0.58064516, 0.51724138]),
 'train_f1': array([0.75239923, 0.76865672, 0.76979742, 0.77372263, 0.75144509,
        0.76102941, 0.76727273, 0.75645756, 0.76383764, 0.75280899])}

In [21]:
## 3가지 모델을 써서 평가를 비교해 보자!
## 이진분류에서 기본적으로 dt, rf, lr 3가지를 가지고 비교해 보자!
model = [DecisionTreeClassifier(), RandomForestClassifier(), LogisticRegression()]
name = ["DT","RF","LR"]

for model, name in zip(model, name):
    print("################## 사용한 갈고리즘", name, "##################")
    for score in ["accuracy","precision","recall","f1"]:
        print(score)
        print("-")
        print(cross_val_score(model, df_tt_x, df_tt_y, scoring=score, cv=10))

################## 사용한 갈고리즘 DT ##################
accuracy
-
[0.68888889 0.64044944 0.66292135 0.61797753 0.64044944 0.75280899
 0.66292135 0.7752809  0.70786517 0.68539326]
precision
-
[0.66666667 0.54545455 0.63157895 0.52173913 0.54545455 0.6875
 0.57692308 0.71875    0.64285714 0.66666667]
recall
-
[0.4        0.35294118 0.32352941 0.38235294 0.41176471 0.64705882
 0.44117647 0.67647059 0.52941176 0.45714286]
f1
-
[0.5        0.42857143 0.42307692 0.4137931  0.42857143 0.66666667
 0.50847458 0.6969697  0.59016393 0.54237288]
################## 사용한 갈고리즘 RF ##################
accuracy
-
[0.71111111 0.60674157 0.66292135 0.65168539 0.6741573  0.71910112
 0.70786517 0.79775281 0.73033708 0.73033708]
precision
-
[0.68       0.44444444 0.63636364 0.55172414 0.60714286 0.62162162
 0.59375    0.70588235 0.66666667 0.65384615]
recall
-
[0.42857143 0.38235294 0.41176471 0.47058824 0.44117647 0.67647059
 0.55882353 0.70588235 0.58823529 0.51428571]
f1
-
[0.51724138 0.4516129  0.46428571 0.491

### 추가교차검증
- 데이터와 분석 목적에 따라서 교차검증을 선택할 수 있다.
- shuffleSplit - 반복 횟수를 훈련 세트, 테스트 세트를 독립적으로 조절해야 하는 경우
- GroupKFold - 데이터 안에 연관된 그룹들이 있을 때
- RepeatedStratifiedKFold, RepeatedKFold - 데이터가 적을 때 안정된 점수를 얻기 위해서 교차검증을 여러번 반복
- Lvo 등 사이킷런에서 찾아보기!

### 필수과제1
- 위에 언급한 추가교차검증 사이킷런에서 제공하는 다양한 교차검증 내용을 리서치하고
- 개념 정리와 코드 정리를 함께 공유해 주세요!
- 최소 3개 이상 정리해 주세요!

### 평가지표 (Evaluation)
- 정확도, 정밀도, 재현율, f1 평가지표
- 성능평가지표는 모델링에 대한 성능 지표, 분류냐 회귀냐에 따라 달라진다.
- 회귀는 실제값과 예측값에 대한 잔차, 오차에 대한 기준으로 스코어링한다.
- 오차값에 절댓값 쓰위 평균 오차를 구하고 RMSE, MSE

- 우리가 배웠던 분류 평가 방법이 4-5개 정도 된다.
- 평가지표는
- 정확도(accuracy)
- 정밀도(precision)
- 재현율(recall)
- f1 스코어
- ROC AUC
- 오차행렬에 대한 내용
- 위의 평가지표는 이진분류 시 사용된다.

In [22]:
df

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone  
0      man        True  NaN  Southampton    no  False  
1    woman       False    C    Cherbourg   yes  False  
2    woman       False  NaN  Southampton   yes   True  
3    woman       False    C  Southampton   yes  False  
4      man        True  NaN  Southampton    no   True  
..     ...         ...  ...          ...   ...    ...  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True  

[891 rows x 15 columns]

### 오차 행렬 confusion matrix
- 4 사분면에 positive 1 / negative 0

        - 예측 클래스
        - N(0) / P(1)
- N (0)
        - TN FP
- 실제 클래스
- P (1)

        - FN TP
- TN은 뭐냐?
- True 실제값과, 예측값의 클래스 같다 True + N 예측값에 대한 N = TN

- TN 예측값은 Neg(0), 실제값 역시 Neg(0) 값
- 이 원리를 그대로 적용해서 FP, FN, TP가 나오는 것

### 정확도
- 예측 결과가 동일한 데이터 개수 / 전체 예측 데이터 개수
- 정확도 계산식
- (TN + TP) / (TN + TP + FN + FP)
- 절대로 정확도만 가지고 평가할 수 없다.
- 정확도 지표만 보면 안 되는 이유?
- 100개 데이터가 있다. 90개는 0, 10개는 1, 0을 예측하는 게 90% 클래스 불균형
- 오차행렬을 만드는 이유가 정확도만 보면 문제가 되어서 진행하는 것이다.

### 정밀도
- TP / (FP + TP)
- 1로 예측한 것들 중 진짜 1인 경우가 얼마나 되는가?
- 예측을 Positive로 한 대상 중에서 실제 값이 Positive로 일치한 데이터
- positive에 대한 성능을 정밀하게 예측한다고 해서 '양성 예측도'라고도 한다.

### 재현율
- 실제 값이 Positive 대상 중에서 예측과 실제 값이 Positive로 일치한 데이터 비율
- TP / (FN + TP)
- 민감도(Sensitivity), TPR(True Positive Rate)라고도 한다.

### 정밀도와 재현율은 Trade off 관계
- 정밀도 - TP가 올라가거나, FP 점수
- TP / (FP + TP)

- 재현율 - FN 점수
- TP / (FN + TP)

- 임계값은 1과 0에 대해서 예측할 때 디폴트 값이 0.5를 기준으로 해서 1이 되거나 0.499 0이 되거나
- 분류를 하는데, 해당 임계값이 조절 가능하다.
- 0.6부터 1로, 0.59 아래는 0으로 조절이 가능하다.